pkg/cursor/cursor.go

```diff
@@ -241,46 +241,22 @@ func (cur *crsr) CurrentPos() records.IteratorPos {
 	return cur.it.CurrentPos()
 }
 
-// Offset moves the cursor position either backward (negative offset) or
-// forward (positive offset) value
+// Offset sets the cursor to backward to tru when offset is negative. when offset is positive, Offsets moves the curson position forward by offset value.
 func (cur *crsr) Offset(ctx context.Context, offs int) {
-	if offs == 0 {
-		return
-	}
-
-	var pos records.IteratorPos
-	pos = records.IteratorPosUnknown
-
-	bkwd := false
-	if offs < 0 {
-		bkwd = true
-		offs = -offs
-		_, _, err := cur.Get(ctx)
-		pos = cur.CurrentPos()
+	switch {
+	case offs < 0:
 		cur.SetBackward(true)
```

Comment:
I afraid your change breaks the logic. The cursor never iterates "backward". It can skip some records back, but it always return record in "forward" way. I am not sure your change will support this behavior.

Reply:
earlier logic as well iterating backward by offset then forwards by limit. i also checked iterator part, iterator supporting backward. Added unit test to verify if cursor backward is working (single partition cursor) if that is working, test is passed. can you take a look at it please ?

Reply:
@adasari I took a look at your PR, but what I still don't understand is:
1. What is wrong with the existing implementation?
2. What is your proposal for the behavior you are trying to change.

Reply:
The current implementation have following challenges (moving our conversation to github)
1. OFFSET=-5000, LIMIT=3, POS=TAIL (when total records < 5000) always returns log entries
- in case of latest records on top scenario, couldn't find end of log entries as negative offset always returns records.
2. pos value in the next page request is incorrect in negative offset case. So pos can't be used in next request. 
  - it is returning the offset post forward movement. 
3. OFFSET=0, LIMIT=1, POS=TAIL is not returning any records
 - This forces us use OFFSET=-10, LIMIT=10, POS=TAIL to get latest 10 log records.

How negative offset/tail scenario is working -
it moves the cursor by offset number to pos value in backward direction and return the limit number of log records by iterating cursor in forward direction.

What is proposed in my changes ?

1. Avoid moving cursor to backward in negative direction by offset and forward direction by limit number of records. only, set the cursor backward to true. negative offset is used to just indicate that cursor should be iterated in backward direction and now offset functionality is different in both head and tail scenario which is same in the current implementation. may need a separate flag/clause to determine head/tail scenarios to keep offset functionality sync ? 

2. returns current pos which can be used in next page request. #2 returning pos after moving the cursor in forward direction. 

3. #3 is not fixed with new proposal, still have to use OFFSET=-10, LIMIT=10, POS=TAIL to get latest 10 log records. May need a separate flag/clause to determine head/tail scenarios ?

Feedback and suggestions welcome. 
Please let me know if you have any questions.

Reply:
@adasari, before answering on your questions, please take into account the following:
a) the iterator works by rules described [here](https://github.com/logrange/range/blob/00aad37c363cdfa6e339f7ea523daa3947c3fde8/pkg/records/chunk/chunkfs/citerator.go#L26)
b) Negative offset shifts current position to the number of records specified in the absolute offset value towards to the head
c) there is no backward iteration, only forward iteration.

Now answering by your questions:
1. According to (a) if the position is negative or out of the range, the first record will be returned.
2. I am not sure that is true. Please take a look [here](https://github.com/logrange/logrange/blob/c38f23eb0cb9d0105c8407f00cf0f7b5981016de/api/client.go#L53) where returned value for the next request is used for iterating over the collection. Also take into account that according (b) and (c) - negative offset just affects the query position, but not its iteration order. Please also be sure, that consecutive  requests don't have offset, otherwise the request position will be affected.
3. That is correct behavior, please refer at (a).

I have to admit, that current iteration implementation is not clear and complicated, but I am not sure it is "incorrect". I am going to simplify it and going to publish new "design" version soon, so you are welcome to suggest for the new implementations.

Reply:
@dspasibenko Thank you for response.

i agree on (a) & (b). regarding (c), iterator supports backward iteration ([0 .. Count)|       ANY       | The record at the pos). the current implementation using backward iteration incase of negative offset [here](https://github.com/logrange/logrange/blob/master/pkg/cursor/cursor.go#L260) . Seems you are misunderstanding my changes. Could you please take a look at cursor_test.go#TestCursorOffset, the tests include both postive and negative offsets.

I will wait for your new design? do you have any ETA or branch where i can test/enhance ? thanks.

Reply:
@adasari when I wrote "there is no backward iteration, only forward iteration." I meant not LQL, nor API supports it. Yes, iterators have the feature, but how this should work using LQL? I understand what you PR does, but I am not sure that it is enough to support the "backward iteration" feature. At least an LQL statement should be introduced. I am not sure it makes sense at the point. 

Regarding another branch - I am going to publish a design doc, will share here for review. Unfortunately this days I am busy with other stuff, but I promise, I will come back to it shortly.